packages/go_router/lib/src/matching.dart

```diff
@@ -188,59 +200,48 @@ List<RouteMatch> _getLocRouteRecursively({
         newParentSubLoc = match.subloc;
       }
 
-      final List<RouteMatch> subRouteMatch = _getLocRouteRecursively(
+      final List<RouteMatch>? subRouteMatch = _getLocRouteRecursively(
         loc: loc,
         restLoc: childRestLoc,
         parentSubloc: newParentSubLoc,
         routes: route.routes,
-        parentFullpath: fullpath,
-        queryParams: queryParams,
-        queryParametersAll: queryParametersAll,
+        pathParameters: subPathParameters,
         extra: extra,
-      ).toList();
+      );
 
       // If there's no sub-route matches, there is no match for this location
-      if (subRouteMatch.isEmpty) {
+      if (subRouteMatch == null) {
         continue;
       }
-      result.add(<RouteMatch>[match, ...subRouteMatch]);
+      result = <RouteMatch>[match, ...subRouteMatch];
     }
     // Should only reach here if there is a match.
-    if (debugGatherAllMatches) {
-      continue;
-    } else {
-      break;
-    }
+    break;
   }
-
-  if (result.isEmpty) {
-    return <RouteMatch>[];
+  if (result != null) {
+    pathParameters.addAll(subPathParameters);
   }
-
-  // If there are multiple routes that match the location, returning the first one.
-  // To make predefined routes to take precedence over dynamic routes eg. '/:id'
-  // consider adding the dynamic route at the end of the routes
-  return result.first;
```

Comment:
This always returns first, might as well use a single object to hold the result. This is probably some left over code from previous refactoring.